### PR TITLE
fix: Consider missing optional input/output artifacts with same name

### LINKED ISSUE #1750: Output artifact in same location as optional input artifact is not saved
**Is this a BUG REPORT or FEATURE REQUEST?**: BUG REPORT

**What happened**: When an output artifact location matches that of an optional input artifact which is not supplied, the output artifact fails to save properly. This is related to #1503 and #1660.

**How to reproduce it (as minimally and precisely as possible)**:
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: optional-input-output-artifact-same-location-
spec:
  entrypoint: plan
  templates:
  - name: plan
    steps:
    - - name: write
        template: file
    - - name: read
        template: file
        arguments:
          artifacts:
          - name: file
            from: "{{steps.write.outputs.artifacts.file}}"
  - name: file
    inputs:
      artifacts:
      - name: file
        path: /tmp/file1.txt
        optional: true
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo 'test' > /tmp/file1.txt"]
    outputs:
      artifacts:
      - name: file
        path: /tmp/file1.txt
```
When this is run, the `write` step passes, but the `read` step fails with this message: `inputs.artifacts.file was not supplied`. If the location of the optional input file is changed to `/tmp/file2.txt`, then both steps pass. This indicates that the output artifact is failing to save properly, but isn't raising an error in the process even though it isn't labeled as optional.

**Environment**: Argo version: 2.4.1

### PR DESCRIPTION
**Pull request by simster7**
Fixes #1750

### CODE REVIEW

File: test/e2e/functional/same-input-output-path-optional.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+apiVersion: argoproj.io/v1alpha1
```
Comment (alexec) [inline test/e2e/functional/same-input-output-path-optional.yaml:1]: move to `../testdata`

File: test/e2e/functional/same-input-output-path-optional.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            arguments:
+              artifacts:
+                - name: file
+                  from: "{{steps.write.outputs.artifacts.file}}"
+    - name: file
+      inputs:
+        artifacts:
+          - name: file
+            path: /tmp/file1.txt
+            optional: true
```
Comment (alexec) [inline test/e2e/functional/same-input-output-path-optional.yaml:25]: I'm not sure optional actually works - can you run without it?

File: test/e2e/functional/same-input-output-path-optional.yaml (hunk 1 above)
Comment (simster7) [reply test/e2e/functional/same-input-output-path-optional.yaml]: Moved!

File: test/e2e/functional/same-input-output-path-optional.yaml (hunk 2 above)
Comment (simster7) [reply test/e2e/functional/same-input-output-path-optional.yaml]: Running without `optional: true` fails the validation as expected:

```
$ argo submit art.yaml
2020/05/13 14:56:52 Failed to submit workflow: templates.plan.steps[0].write templates.file inputs.artifacts.file was not supplied
```

**Comment by alexec** [comment]
Review alongside #3019.

**Comment by simster7** [comment]
> Conditional approval - can you check the execution time of the e2e tests and make sure they have increased significantly?

Assuming you meant "make sure that they _haven't_ increased significantly". If so, the test only adds around 8-10 seconds. I think this is good enough.

> Review alongside #3019.

I took a look and it looks orthogonal.

**Comment by simster7** [comment]
Back-ported to v2.8.1

**Pull request merged**